hpke.go

```diff
@@ -184,21 +184,38 @@ func keySchedule(suite CipherSuite, mode Mode, sharedSecret, info, psk, pskID []
 	return params, nil
 }
 
+// Context represents an HPKE context encoded on the wire.
+type Context struct {
+	Role                 Role
+	KemId, KdfId, AeadId uint16
+	ExporterSecret       []byte `tls:"head=1"`
+	Key                  []byte `tls:"head=1"`
+	Nonce                []byte `tls:"head=1"`
+	Seq                  uint64
+}
+
+// Role specifies the role of a party in possession of a Context: if equal to
+// `0`, then the party is the sender; if equal to `0`, then the party is the
+// receiver.
+type Role uint8
+
+const (
+	roleSender   Role = 0x00
+	roleReceiver Role = 0x01
+)
+
 type cipherContext struct {
-	key            []byte
-	nonce          []byte
-	exporterSecret []byte
-	aead           cipher.AEAD
-	seq            uint64
-	suite          CipherSuite
+	inner Context
```

Comment:
I don't love this; it feels like we're letting out wire format infect our in-memory format.  I might prefer to construct `Context` objects at marshal/unmarshal time, and copy individual fields over.

Reply:
This would also save you all of the `s/ctx./ctx.inner./` changes below.

Reply:
IMO it's cleaner to define each value precisely once. Imagine we had two structs: 

```
type cipherContext struct {                                                                                                                                                                                        
    key            []byte
    nonce          []byte
    exporterSecret []byte
    aead           cipher.AEAD
    seq            uint64
    suite          CipherSuite
    // ... [truncated] ...
}
```
and
```
// context represents an HPKE context encoded on the wire.
type context struct {
    Role           contextRole
    KEMID          KEMID
    KDFID          KDFID
    AEADID         AEADID
    ExporterSecret []byte `tls:"head=1"`
    Key            []byte `tls:"head=1"`
    Nonce          []byte `tls:"head=1"`
    Seq            uint64
}   
``` 
If we wanted to change the name of `nonce`, say, then we would need to change it in two places. Better to name it once, I think.

Reply:
If you want to preserve that property, then just hoist all your fields in `context` up into `cipherContext` and slap a `tls:"omit"` on the things you don't want marshaled.

```
type cipherContext struct {
	// Marshaled fields
	Role           contextRole
	KEMID          KEMID
	KDFID          KDFID
	AEADID         AEADID
	ExporterSecret []byte `tls:"head=1"`
	Key            []byte `tls:"head=1"`
	Nonce          []byte `tls:"head=1"`
	Seq            uint64

	// Operational structures
	AEAD  cipher.AEAD `tls:"omit"`
	Suite CipherSuite `tls:"omit"`

	// Historical record
	Nonces        [][]byte          `tls:"omit"`
	SetupParams   setupParameters   `tls:"omit"`
	ContextParams contextParameters `tls:"omit"`
}
```

Reply:
That works for me! Do you prefer "cipherContext" to "context" for the name of this struct?

Reply:
I personally prefer `context`, since it's more than just the bits needed for encryption/decryption.

Reply:
Yeah, plain `context` works for me, not sure why it was ever `cipherContext`.  Probably @chris-wood 's fault :)

Reply:
Done.

Reply:
Hah! I'm happy to take the blame and beating. It usually is my fault.